### PR TITLE
Extract JsonWriter from JsonModelWriter

### PR DESCRIPTION
**Pull request by reinsch82**
Remove unused code

Extract BuildNameHandler

This is a precursor to https://github.com/gradle/gradle/pull/29787

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.apache.groovy.json.internal.CharBuf
+import java.io.Writer
+
+open class JsonModelWriterCommon(val writer: Writer) {
+    // This is a map of booleans that indicate whether the next element in the array should be preceded by a comma.
+    var commaContext = ArrayDeque<Boolean>()
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:24]: 🤔  Consider having one boolean variable instead.

It's set to false whenever the level is increased, and it's set to true whenever the level is decreased.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    private fun increaseLevel() {
+        commaContext.addLast(false)
+    }
+
+    private fun decreaseLevel() {
+        commaContext.removeLast()
+    }
+
+    fun endArray() {
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:67]: 💅  We should move all public methods to the top of the class, and arrange them logically

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt [hunk 3]
```diff
@@ -37,15 +35,11 @@ enum class DiagnosticKind {
     INCOMPATIBLE_TASK
 }
 
-
-class JsonModelWriter(val writer: Writer) {
+class JsonModelWriter(writer: Writer): JsonModelWriterCommon(writer) {
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:38]: ❌  Since the new writer is a general utility to write JSON's, it should be used via composition rather than inheritance

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt [hunk 4]
```diff
@@ -55,49 +49,39 @@ class JsonModelWriter(val writer: Writer) {
 
     fun endModel(details: ProblemReportDetails) = with(details) {
         endArray()
-
         comma()
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:52]: ❓  Is this comma required here?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt [hunk 5]
```diff
@@ -116,7 +100,6 @@ class JsonModelWriter(val writer: Writer) {
                 }
 
                 if (parts != null) {
-                    if (summary != null) comma()
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:119]: 🎉  It's nice not to have to care about the separator anymore!

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.gradle.api.initialization.Settings
+import org.gradle.api.internal.SettingsInternal
+import org.gradle.internal.InternalBuildAdapter
+
+class BuildNameHandler : InternalBuildAdapter() {
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt:23]: ❓  What is the reason we want to extract this?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt (hunk 4 above)
Comment (reinsch82) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:52]: yes, it breaks otherwise

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt (hunk 6 above)
Comment (reinsch82) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt:23]: I'm reusing this for problems report generation

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt (hunk 4 above)
Comment (alllex) [reply platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriter.kt:52]: This is rather strange, because the following `property()` call should insert a comma automatically, given that we call `endArray()` and decrease a level.

Should we add some tests for the Json writing utility?

**Review by donat — COMMENTED** [review]
I'm fine with the changes given that you address @alllex's review items.

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt (hunk 6 above)
Comment (alllex) [reply platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameHandler.kt:23]: Then I think it should be turned into an independent service that provides the build name once it is available. 

We should avoid having classes that need manual lifecycle management if their registration is unconditional.

At the same time, there are two approaches to this now: explicit and implicit `ListenerManager` registration. We don't have a guideline on which one is a best practice.

I think you should go with the explicit one for now. See an example here:
https://github.com/gradle/gradle/blob/94faa3d523951f395123f459d4f513bd417a654f/platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinCompilerContextDisposer.kt

Given it becomes a service, I'd
- Change its name to something more appropriate like `BuildNameProvider`
- Add a javadoc
- Hide the `bulidName` field and throw an exception in case the build name is requested before it was initialized

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

File: build-logic/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt [hunk 7]
```diff
@@ -114,8 +114,10 @@ abstract class ExternalModulesExtension(isBundleGroovy4: Boolean) {
     val junitPlatformEngine = "org.junit.platform:junit-platform-engine"
     val jzlib = "com.jcraft:jzlib"
     val kotlinCompilerEmbeddable = futureKotlin("compiler-embeddable")
+    val kotlinxSerializationJson = "org.jetbrains.kotlinx:kotlinx-serialization-json"
     val kotlinReflect = futureKotlin("reflect")
     val kotlinStdlib = futureKotlin("stdlib")
+    val kotlinxSerializationCore = "org.jetbrains.kotlinx:kotlinx-serialization-core"
```
Comment (alllex) [inline build-logic/dependency-modules/src/main/kotlin/gradlebuild/modules/extension/ExternalModulesExtension.kt:120]: 💅  Use alphabetical order

File: gradle.properties [hunk 8]
```diff
@@ -1,5 +1,5 @@
 # Reduce Xmx after https://github.com/gradle/gradle-private/issues/4168 is resolved
-org.gradle.jvmargs=-Xmx3000m -XX:MaxMetaspaceSize=768m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
+org.gradle.jvmargs=-Xmx3300m -XX:MaxMetaspaceSize=768m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
```
Comment (alllex) [inline gradle.properties:2]: 🤔  Is this necessary for this PR?

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt [hunk 9]
```diff
@@ -44,10 +44,12 @@ import org.gradle.internal.cc.impl.initialization.DefaultConfigurationCacheProbl
 import org.gradle.internal.cc.impl.initialization.InstrumentedExecutionAccessListenerRegistry
 import org.gradle.internal.cc.impl.initialization.VintageInjectedClasspathInstrumentationStrategy
 import org.gradle.internal.cc.impl.models.DefaultToolingModelParameterCarrierFactory
+import org.gradle.internal.cc.impl.problems.BuildNameProvider
 import org.gradle.internal.cc.impl.problems.ConfigurationCacheProblems
 import org.gradle.internal.cc.impl.services.ConfigurationCacheBuildTreeModelSideEffectExecutor
 import org.gradle.internal.cc.impl.services.VintageEnvironmentChangeTracker
 import org.gradle.internal.configuration.problems.DefaultProblemFactory
+import org.gradle.internal.event.ListenerManager
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultBuildTreeModelControllerServices.kt:52]: 🤔   This import does not seem to be used by other changes in the file

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import org.gradle.internal.InternalBuildAdapter
+import org.gradle.internal.concurrent.Stoppable
+import org.gradle.internal.event.ListenerManager
+import org.gradle.internal.service.scopes.Scope
+import org.gradle.internal.service.scopes.ServiceScope
+
+/**
+ * A utility class that provides the name of the build.
+ */
+
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt:30]: 💅  Empty line

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+@ServiceScope(Scope.BuildTree::class)
+class BuildNameProvider(val listenerManager: ListenerManager) : InternalBuildAdapter(), Stoppable {
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt:32]: The member should be private

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+    init {
+        listenerManager.addListener(this)
+    }
+
+    override fun stop() {
+        listenerManager.removeListener(this)
+    }
+
+    var buildName: String? = null
+        get() = field
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt:43]: ❌  This should not be publicly mutable.

We also should throw an exception in case someone tries to get the name before it is available.

It seems that the root project name is always available. So the type should not be nullable.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/CommonReport.kt [hunk 13]
```diff
@@ -126,7 +127,13 @@ class ConfigurationCacheReport(
             val hashingStream = HashingOutputStream(Hashing.md5(), spoolFile.outputStream().buffered())
 
             private
-            val writer = HtmlReportWriter(hashingStream.writer(), htmlReportTemplate)
+            val hashingWriter: Writer = hashingStream.writer()
+
+            private
+            val jsonModelWriter: JsonModelWriter = JsonModelWriter(JsonModelWriterCommon(hashingWriter))
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/CommonReport.kt:133]: 🤔  We don't seem to need these fields in the class itself, and only need those values to create the final writer.
Should we create a function that creates the `HtmlReportWriter`?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+        }
+    }
+
+    private
+    fun elementSeparator() {
+            if (commaContext) {
+                comma()
+            } else {
+                commaContext = true
+            }
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:93]: 💅 Formatting

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.apache.groovy.json.internal.CharBuf
+import java.io.Writer
+
+open class JsonModelWriterCommon(val writer: Writer) {
+    // This is a map of booleans that indicate whether the next element in the array should be preceded by a comma.
+    private var commaContext = false
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:24]: ❌  The comment is outdated, and I don't one is required.

I'd also suggest to rename the field to a more telling `nextItemNeedsComma` or something along those lines. It's also not only about arrays, but properties within an object.

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.cc.impl.problems
+
+import org.apache.groovy.json.internal.CharBuf
+import java.io.Writer
+
+open class JsonModelWriterCommon(val writer: Writer) {
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:22]: 🤔  The class probably does not need to be `open` anymore.
It is now also a handy general-purpose JSON writer. Let's simplify the name. How about `JsonWriter`?

File: platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+        write('[')
+    }
+
+    fun endArray() {
+        write(']')
+        decreaseLevel()
+    }
+
+
+    fun comma() {
```
Comment (alllex) [inline platforms/core-configuration/configuration-problems-base/src/main/kotlin/org/gradle/internal/cc/impl/problems/JsonModelWriterCommon.kt:54]: 💅  An extra empty line. Also, does this method need to be public?

File: subprojects/distributions-dependencies/build.gradle.kts [hunk 18]
```diff
@@ -191,6 +191,8 @@ dependencies {
         api(libs.jtar)                  { version { strictly("2.3") }}
         api(libs.kotlinCoroutines)      { version { strictly("1.5.2") }}
         api(libs.kotlinCoroutinesDebug) { version { strictly("1.5.2") }}
+        api(libs.kotlinxSerializationCore)   { version { strictly("1.6.3") }}
+        api(libs.kotlinxSerializationJson)            { version { strictly("1.6.3") }}
```
Comment (alllex) [inline subprojects/distributions-dependencies/build.gradle.kts:195]: 💅  Formatting

File: subprojects/distributions-dependencies/build.gradle.kts [hunk 19]
```diff
@@ -191,6 +191,8 @@ dependencies {
         api(libs.jtar)                  { version { strictly("2.3") }}
         api(libs.kotlinCoroutines)      { version { strictly("1.5.2") }}
         api(libs.kotlinCoroutinesDebug) { version { strictly("1.5.2") }}
+        api(libs.kotlinxSerializationCore)   { version { strictly("1.6.3") }}
```
Comment (alllex) [inline subprojects/distributions-dependencies/build.gradle.kts:194]: ⚠️  We have updated the dependency version here. I am not sure if this was intended.
These dependencies are also unrelated to any of the JSON writer logic we are refactoring, right?

Comment (reinsch82) [reply subprojects/distributions-dependencies/build.gradle.kts:194]: That is true, I ended up changing it, because I thought of using it for the `JsonWriter`.

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt [hunk 20]
```diff
@@ -39,12 +38,19 @@ class BuildNameProvider(private val listenerManager: ListenerManager) : Internal
         listenerManager.removeListener(this)
     }
 
-    var buildName: String? = null
-        get() = field
+    private
+    lateinit var rootBuildName: String
+
+    fun buildName(): String {
+        if (this::rootBuildName.isInitialized) {
```
Comment (alllex) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/BuildNameProvider.kt:45]: ❓  Is there a reflection cost for this?

I think having a nullable private field and checking for null should be enough here.
We can also do the following to have the right nullability expectations:
```
rootBuildName = settings.rootProject.name!!
```

**Comment by reinsch82** [comment]
@bot-gradle test this

**Comment by reinsch82** [comment]
@bot-gradle test this

**Review by alllex — APPROVED** [review]
🎉

**Comment by reinsch82** [comment]
@bot-gradle test this
